src/utils/spark.py

```diff
@@ -0,0 +1,223 @@
+            self.ipts_password = file.read().strip()
+
+        # Static arguments/params applied to jobs using this session
+        self.fetch_size = "10000"
+        self.initial_compression = "snappy"
+        self.final_compression = "zstd"
+
+        self.spark = SparkSession.builder.appName(self.app_name).getOrCreate()
+
+
+class SparkJob:
+    """
+    Class to manage Spark jobs for reading, repartitioning, and uploading data.
+    Each job corresponds to a single iasWorld table.
+
+    Attributes:
+        session: The shared Spark session containing the Spark connection and
+            database credentials.
+        table_name: The name of the iasWorld table to read from. Should be
+            predicated with 'iasworld.'.
+        taxyr: The tax year(s) to filter and partition by.
+        cur: The cur values to filter and partition by.
+        predicates: A list of SQL predicates for chunking JDBC reads.
+        use_partitions: Whether to partition the output data by taxyr and cur.
+        initial_dir: The initial directory to write the data to, relative to
+            the Docker container.
+        final_dir: The final directory to write the repartitioned data to,
+            relative to the Docker container.
+    """
+
+    def __init__(
+        self,
+        session: SharedSparkSession,
+        table_name: str,
+        taxyr: int | list[int] | None,
+        cur: str | list[str] | None,
+        predicates: list[str] | None,
+        use_partitions: bool,
+        initial_dir: str,
+        final_dir: str,
+    ) -> None:
+        self.session = session
+        self.table_name = table_name
+        self.taxyr = taxyr
+        self.cur = cur
+        self.predicates = predicates
+        self.use_partitions = use_partitions
+        self.initial_dir = (
+            (Path(initial_dir) / strip_table_prefix(self.table_name))
+            .resolve()
+            .as_posix()
+        )
+        self.final_dir = (
+            (Path(final_dir) / strip_table_prefix(self.table_name))
+            .resolve()
+            .as_posix()
+        )
+
+        # Both of these must be set to avoid situations where we create
+        # partitions by taxyr but not cur, and visa-versa
+        if self.use_partitions and (
+            (self.taxyr is None) != (self.cur is None)
+        ):
+            raise ValueError(
+                (
+                    f"Error for table {self.table_name}: "
+                    "both 'taxyr' and 'cur' must be set if one is set."
+                )
+            )
+
+    def get_filter(self) -> str | None:
+        """
+        Translates the taxyr and cur values into SQL used to filter/limit the
+        values read from the table.
+        """
+        if self.taxyr is None:
+            return None
+
+        if isinstance(self.taxyr, list):
+            filter = f"taxyr IN ({', '.join(map(str, self.taxyr))})"
+        else:
+            filter = f"taxyr = {self.taxyr}"
```

Comment:
**[Question, non-blocking]** Maybe i'm missing something, but aren't we already filtering by `taxyr` via the SQL expressions that comprise the predicates matrix?

Reply:
I simplified the filter setup a bit in 4090211. Like partitions, the filter is now just constructed from any passed `taxyr` and `cur` values. I also removed the interaction of predicates and `taxyr`.